### ISSUE TITLE
Research on existing performance testing framework

### ISSUE THREAD

**Issue by dextermallo** [dextermallo/GSoC-2023#3]
Research on existing performance testing framework.

REF: https://owasp.slack.com/archives/C03EXFGM4FJ/p1683814881754659

**Comment by dexter-ys** [comment]
There are several **Framework for Testing WAF** (a.k.a FTW):
- [coreruleset/ftw](https://github.com/coreruleset/ftw)
- [fastly/wafefficacy](https://github.com/fastly/wafefficacy) [[1]](https://www.fastly.com/blog/the-waf-efficacy-framework-measuring-the-effectiveness-of-your-waf)
- [microsoft/WAFBench](https://github.com/microsoft/WAFBench)
- [digitalwave/ftwrunner](https://github.com/digitalwave/ftwrunner)
---

**References**

[1. The WAF efficacy framework: measuring the effectiveness of your WAF | Fastly](https://www.fastly.com/blog/the-waf-efficacy-framework-measuring-the-effectiveness-of-your-waf)

**Comment by dexter-ys** [comment]
Here's a framework draft - profiling based on the logs yield from the services (which can be mounted). In this way, the analysis can be based on the server without any interruption to the server. And it should not have a compatibility issue as long as the service supports logging the matrix we want.

## Use case

When a contributor wants to test the possible performance impact caused by changes in the rules or regex, they will:
1. run the WAF and go-ftw.
2. run the script to generate a performance report (report A)
3. doing code changes...
4. repeat 2. and 3. (report B)
5. run another script to get a benchmark report (based on reports A and B)

## Matrices

These matrices should be able to capture (discuss in further meetings for must-have/optional)
- Functional
  - Error rate (We can compare before/after the test case pass percentage. Also, we can use modsec_audit.log to compare the diff and see what happens)
  - Connection Time (by [adding configuration in nginx](https://stackoverflow.com/questions/69534518/what-does-connection-time-mean-in-nginx))
  - RTT (currently in stdout)
  - Concurrency Maximum (use utils such as [ab](https://httpd.apache.org/docs/2.4/programs/ab.html), see Fig 1)
- Hardware
  - Resource Utilisation (use `docker stats` with a script. By the timestamp, we can map to the logs and see if there is any anomaly resource utilisation issue. See Fig 2)
  - Network I/O (ditto)


<img width="1552" alt="image" src="https://github.com/dextermallo/GSoC-2023/assets/29012607/0a4ba1d8-81ba-4c36-9314-d53adaa13c33">
(Fig 1. Load test on the proxy via ab)

<img width="1440" alt="docker stats customise" src="https://github.com/dextermallo/GSoC-2023/assets/29012607/eab88562-47ff-4823-ab75-62a1c595caad">
(Fig 2. logs of docker stats) 

```sh
#!/bin/bash
# REF: https://stackoverflow.com/questions/67174376/how-to-print-a-timestamp-with-the-docker-stats
# `tee` is removed for MacOS

now=$(date +'%d_%b_%Y_%H_%M')

while true;
    do printf "\n$(date +'%d_%b_%Y_%H_%M_%S'):\n" >> stats_$now.txt;
    docker stats modsec3-nginx --no-stream >> stats_$now.txt; sleep 1;
done
```



For the hardware matrix, it's possible to get the information via `docker stats`. However, the granularity (can't change to milli-sec and logs all the processes instead of nginx independently) is not quite well. As for the functional matrix, it should be available via the built-in logs from nginx/apache.

## Benefits

- Less or no need to change any source code. (needs to check on existing log settings)
- There are no concerns about the integration so far.
- Able to apply to ModSecurity 2.9, 3.0 and Coraza

## Challenges

1. The log file is huge and time-consuming (If running all test cases). In total, we may need 10+ min and 260+ MB (run twice and without running the script to yield the report). Evaluation based on my run:
- Apache (3517 test cases, skipped 4):
  - Time: 3'56
  - Size: access.log (1.1 MB), error.log (21.9 MB), and modsec_audit.log (53.3 MB)
- Nginx (3517 test cases, skipped 4):
  - Time: 1'26
  - Size: access.log (1.2 MB), error.log (22.7 MB), and modsec_audit.log (35.5 MB)

2. Accuracy might be an issue since we rely on the logs. If the log is inaccurate, the result might be the same.

**Comment by dexter-ys** [comment]
Another two thoughts:

1. Doing the benchmark on the WAF container directly (In-container solution): It is possible to do so, such as by adding a plugin to collect the information when a specific rule is run. However, based on the use case, the approach is inferior to profiling via log analysis.

2. Use a client container to benchmark the WAF container (container-to-container solution): Hosting another container might be a good idea. This way, functional data can be collected whereas the hardware information can be gathered via other monitoring utils, such as [cAdvisior](https://github.com/google/cadvisor).

(cAdvisor might have better visibility to monitor resource utilisation than `docker ps`. I am looking for documentation to see if it is possible to output as a log so it can work with other log files)

<img width="1440" alt="image" style="text-align: center;" src="https://github.com/dextermallo/GSoC-2023/assets/29012607/9d91d3b6-6756-4da9-b11d-e43633f5a442">
<p style="text-align: center;">(Fig 1. example of cAdvisor monitoring on nginx reverse-proxy)</p>

**Comment by fzipi** [comment]
Logging depends a lot on the server. We might use it as a basis for now, but you will be getting information from the web server only, right?

**Comment by fzipi** [comment]
Another option to get metrics from the containers is prometheus. Using https://docs.docker.com/config/daemon/prometheus/ you can get an idea. But maybe this is too restrictive: you need to have docker, and the time used for scraping data needs to be really small...

@dune73 can you provide more insights on the logs?

**Comment by dexter-ys** [comment]
> Logging depends a lot on the server. We might use it as a basis for now, but you will be getting information from the web server only, right?

Does the webserver here mean the WAF? If so, yes we may only retrieve server data here. 
Nonetheless, we can also do both sides, such as another container which serves as a client to perform different test cases to get other matrices like concurrency maximum or RTT.

**Comment by dune73** [comment]
This looks good. I can not chime in on the container discussion, but @fzipi will have you covered there.

As for the logs: You can do micosecond logging with ModSec and the debug-log will output that. But debug-logging is a performance hog and different versions write in different log formats. So don't do that at least not in the basic iteration.

I think staying outside the server is much better. What I think is worthwhile is to limit go-ftw calls to individual rules, but executing them n times. No need to run the entire test suite to determine the impact of a change of a single rule. But running said rule 1K iterations will highlight any performance change. Apparently this moves away from the idea to integrate the load testing into the unit tests when doing CI/CD, but I think that should be a 2nd thought. First we need to be able to measure and get decent results, then we think about integrating in day to day operation.

Important thought: Whenever I did load testing, I had to find out the hard way that the client was not up to the task. So any performance problems I saw were first and foremost the client not keeping up with the server. Please make sure you do not fall into that trap. Extensive testing useful.

If you do as described by you above, you will provide us with a useful base and a framework we can work with. Not super-granular but a good base. When later we could move closer to the server, its logs or the kernel inspecting the server.

**Comment by dexter-ys** [comment]
Hi @dune73, thanks for the comments!

> But debug-logging is a performance hog, and different versions write in different log formats.

Could you please explain more regarding the performance hog? As for the latter, yes it will be a problem, thanks for pointing it out!

> What I think is worthwhile is to limit go-ftw calls to individual rules, but executing them n times.

I just did a PoC to make sure I got the point, I tested with 2 approaches:
1. Code changes in go-ftw which calculates avg time of RTT/Stage.
2. A external script that runs go-ftw and counts avg time of RTT/Stage. (since in previous meeting mentioned that avoiding code changes on the existing package would be much more ideal)

Here are the results (based on one test case `932115-1` in a test suite). Before/After represents a regex change performed on the SecRule (from `@rx` to `!@rx`.)

<img width="1048" alt="image" src="https://github.com/dextermallo/GSoC-2023/assets/29012607/6e86e849-c149-457f-b0c0-be15e98a1556">

Code change for references:

```go
// cmd/run.go/rumCmd:53
// support single file
var files string
files = fmt.Sprintf("%s/**/*.yaml", dir)

if strings.HasSuffix(dir, ".yaml") || strings.HasSuffix(dir, ".yml") {
	files = dir
} else {
	files = fmt.Sprintf("%s/**/*.yaml", dir)
}
```

```go
// runner/run.go:23
var GlobalRTTTotal int64 = 0
var GlobalStageTotal int64 = 0

const ITERATIONS int = 1


// run.go/RunTest:97
for _, stage := range testCase.Stages {
	ftwCheck := check.NewCheck(runContext.Config)

	for i := 0; i < ITERATIONS; i++ {
		if err := RunStage(runContext, ftwCheck, testCase, stage.Stage); err != nil {
			return err
		}
	}

	avgRTT := float64(GlobalRTTTotal/int64(ITERATIONS)) / 1e6
	avgStage := float64(GlobalStageTotal/int64(ITERATIONS)) / 1e6
	fmt.Printf("Iteration: %d\nAverage RTT: %fms\nAverage Stage Time: %fms\n", ITERATIONS, avgRTT, avgStage)
}

// run.go/RunStage:184
GlobalRTTTotal += roundTripTime.Nanoseconds()
GlobalStageTotal += stageTime.Nanoseconds()
```

```sh
# Approach 2
#!/bin/bash

total_time=0
iterations=1000
test_yaml=<yaml_path>/.../REQUEST-932-APPLICATION-ATTACK-RCE/932115.yaml

for ((i=0; i<$iterations; i++))
do
    start_time=$(gdate +%s.%6N)    
    result=$(go run ./main.go run -d $test_yaml)
    end_time=$(gdate +%s.%6N)

    echo "$result" | awk -F': ✔ passed in |ms' '{print $2}' | awk '{print $1}' | sed '/^\s*$/d' >> _stage_time.txt
    echo "$result" | awk -F'\\(RTT |ms\\)' '{print $2}' | sed '/^\s*$/d' >> _rtt_time.txt
    execution_time=$(echo "$end_time - $start_time" | bc)
    total_time=$(echo "$total_time + $execution_time" | bc)
done

total_rtt=$(awk '{ sum += $1 } END { printf "%.6f\n", sum }' _rtt_time.txt)
avg_rtt=$(echo "scale=6; $total_rtt/$iterations" | bc)
total_stage_tme=$(awk '{ sum += $1 } END { printf "%.6f\n", sum }' _stage_time.txt)
avg_stage_time=$(echo "scale=6; $total_stage_tme/$iterations" | bc)

echo "Average RTT: $avg_rtt ms"
echo "Average Stage Time: $avg_stage_time ms"
echo "Total execution time: $total_time seconds"

rm -rf _rtt_time.txt _stage_time.txt
```

> Important thought: ...

I remembered there is a similar ticket for this issue. Does it make sense with this logic:
```ts
let timeout = 30;

// where checkStatus(server) pings the health-check endpoint or uses 
// other methods to check if the server is up
while (!checkStatus(server) && timeout--) {
    sleep(10);
}

runTest();
```

**Comment by dexter-ys** [comment]
> Previous Draft: [#3](https://github.com/dextermallo/GSoC-2023/issues/3#issuecomment-1566084347)

# Define the problem

Quoted from the [original proposal](https://owasp.org/www-community/initiatives/gsoc/gsoc2023ideas): 

"... Experience shows you do not really know the performance of a rule until you have tried it out. If you want to test it against a variety of payloads it’s quite a lot of manual work."

"...So the idea is to design a facility (typically a docker container) that is being configured with a rule and then used to test this rule with a variety of payloads and returns a standard report about the performance of the rule."

"Bonus points if multiple engines are covered (ModSecurity 2, libModSecurity 3, Coraza, …)"

Breaking down, we aim to fulfil the following:
- **Precision and accuracy**: One of the concerns about performance testing is that the test itself also affects the performance. Therefore, we need to reduce the performance impact of the testing framework as much as possible. Specifically, considering from client-server model, the server-side log is desired because the network latency and other friction (e.g., computational resource waste) are ignored.
- **Test with different payloads**: We need to test with different payloads to see if the rule works as expected.
- **Compatibility**: If feasible, the testing framework should be compatible with different engines.

# Is there any prior work done on this?

We have reviewed some works in [#3](https://github.com/dextermallo/GSoC-2023/issues/3).
Some literature to see what kinds of matrices can be utilised for testing in [#5](https://github.com/dextermallo/GSoC-2023/issues/5)

# Solution

## Sidecar Pattern

[Sidecar pattern](https://learn.microsoft.com/en-us/azure/architecture/patterns/sidecar), as the name suggests, is a pattern that adds a sidecar container to the main container. The sidecar container is responsible for a specific task, such as logging, monitoring, etc. 

Utilising sidecar patterns for performance tests has some advantages:
- **Isolation**: Sidecar pattern isolates performance-related tasks from the main application. This isolation helps ensure that the performance of the sidecar components does not impact the primary application's performance.
- **Observability**: We can easily integrate monitoring and logging functionalities without modifying the main application's code in a sidecar pattern. Therefore, we can collect performance metrics and analyze them separately from the primary application, providing valuable insights into its behaviour under various load conditions.
- **Modularity**: The sidecar pattern promotes modularity by decoupling functionalities into separate components, which is easier to update or replace specific functionalities without affecting the entire application.

In this case, we can utilise the sidecar pattern to test the performance of the WAF. The merits contributed by the sidecar pattern help to achieve the goals (precision and accuracy, testing with different payloads, compatibility) of this project.

## Architecture

![Architecture diagram](https://github.com/dextermallo/GSoC-2023/assets/29012607/38536fae-b837-4ff2-8493-4e7aa6fe4fc1)


The figure depicts an overview of adapting sidecar patterns into performance testing. In this architecture, the original service remains unchanged (so it is totally safe for the original use, which is from the client to the server), but we can evaluate performance.

Here are explanations of each component:

- **Fuzz Proxy**: the fuzzy proxy is responsible for fuzzing payloads and sending them to the WAF. For example:
    ```sh
    # assume we are testing this rule
    SecRule REQUEST_METHOD "@rx ^(?:GET|HEAD)$"

    # fuzz proxy can help us to add random payloads for two purposes:
    # 1. Adding/Changing/Removing payloads will affect the result
    # For instance, in this case, if we use POST instead of GET, it will become a false positive

    # 2. Adding/Changing/Removing payloads which will not affect the result
    # In this case, if we add a header "Content-type"="xml"
    # It will not affect the result
    ```
- **log analysis script (one-time log analysis)**: the script is used to parse the logs yielded from the server. It only requires to execute once (twice for before/after comparison).
- **ElasticSearch/Prometheus**: ElasticSearch and Prometheus are used to collect hardware matrices. They are used to collect hardware matrices such as CPU, memory, network I/O, etc. Commonly, they require the service to be instrumented with their client library (e.g., HTTP endpoints). 
- **cAdvisor**: cAdvisor is used to collect hardware matrices. It is a container that collects, aggregates, processes, and exports information about running containers. It is used to collect hardware matrices such as CPU, memory, network I/O, etc. (privileged mode is required)

## Use case

As a contributor, I want to test code changes for a regex in the WAF. For example:

```sh
# original
SecRule REQUEST_METHOD "@rx ^(?:GET|HEAD)$" \

# after a code change
SecRule REQUEST_METHOD "@rx ^(?:GET|POST|HEAD)$" \
```

They can do a performance test with the following steps:

1. Run the WAF (e.g., docker-compose.yaml)
2. Run the sidecar containers with another docker-compose.yaml. 
3. Run a load test (e.g., using go-ftw and specify yaml file)
4. Run the script to generate a performance report (before-report)
5. doing code changes...
6. repeat 2, 3, and 4 to generate another performance report (after-report)
7. compare before-report and after-report to see the benchmark. (e.g., another script)

## Matrices

These matrices are available:
- **Functional** (from log analysis)
    - **Error rate**: We can compare before/after the test case pass percentage. Also, we can use `modsec_audit.log` to compare the diff and see what happens
    - **Connection Time**: by [adding configuration in nginx](https://stackoverflow.com/questions/69534518/what-does-connection-time-mean-in-nginx)
    - **RTT**: currently in stdout
    - **Concurrency Maximum**: use utils such as [ab](https://httpd.apache.org/docs/2.4/programs/ab.html)
    - average RTT, average top 10% RTT, average last 10% RTT, etc: ditto
- **Hardware** (from ElasticSearch, Prometheus, and cAdvisor)
    - **Resource Utilisation**
    - **Network I/O**

## Limitations

- **Log formats**: Different WAFs have different log formats. Therefore, we need to write different parsers for different WAFs.
- **Evaluation**: Although the performance test is based on comparing before/after a code change, assume one of them (before or after) is captured incorrectly, the result is not accurate as well. We should find a way to ensure the logs are reliable. (Or, assuming it is a production environment, we can test with different log level)
- **False positive**: When performing fuzz testing, it is critical that we define a minimal scope for the random part. Otherwise, It is hard to evaluate whether the test is legit. For example:
    ```sh
    # assume we are testing this rule
    SecRule REQUEST_METHOD "@rx ^(?:GET|HEAD)$"

    # Therefore, the true positive is GET request and the false positive is POST request
    # In this case, we can perform fuzz testing with GET and POST request
    ```

**Comment by dune73** [comment]
Thank you for these extensive considerations. I had planned to take
part in the meeting later today, but I overlooked a meeting in school
of my younger son. I will therefore give you my feedback in writing.

Not sure I responded on the performance hog question. The debug log
can easily write 5-10K lines of log for a single GET request.
It's quite cool when dealing with ModSec 2.9 performance problems,
but it's manual labor and the ModSec 3 debug log is less useful
due to the absense of certain performance indicators inside the
log.

I would rather not use it, as it's very specific to various
versions and I am sure it slows down a server.

Your results testing 932115-1 look interesting. I am not entirely
sure I can make a call on this, but if you feel the subtle differences
are enough and they are consistent, then I think you are on good tracks
with one of these methods. What I think a bit troubling is that the
two methods bring differring results. go-ftw got faster, the
external script got slower. I think you need to make a substantial
change to the regex. A mere negation is trivial for the engine
and could lead you to run after a red herring.

Code: What is gdate? Gnu version of date? GO version of date?

Other than that you make proficient use of standard shell tools
and bash. This code looks good. Nice to see young people use
awk and bc.

"Important thought: ..." - You might have gotten me wrong here.
The problem is not that the server is not available. The problem
is that - in my experience - in any HTTP transaction the client 
uses far more resources than the server. So even if everything
was deterministic, subtle changes in the environment of the
client have a huge effect that hide the performance differences
on the server side.

Definition of the problem: Very good from my perspective.

I like the idea around the sidecar architecture. I can not 
really comment on the container use. What is unclear to me
is the role and need for cAdvisor when compared to ES / Prometheus.
Are these options / alternatives?

Fuzzying: I see the need for diverse test cases, but I am
not sure fuzzying of payloads is the right approach and 
if you get consistent results by introducing small changes
into the payload via fuzzing.

One problem with fuzzing I see (I think you saw it too) is
the different rules hitting, thus changing the output (and
therefore also changing the resources needed to write the
output). Very hard to use the resulting numbers for
calculations.

Maybe it helps if you explain your reasoning for fuzzing. 
Maybe to Felipe in the call.

Use case: Makes perfect sense.

Matrices: Don't use the modsec_audit.log. It's far too hard to
parse and you do not need this level of information. Going
with the error log is all you need (-> naked alerts without the
rest of the request and response).

Connection Time: This is an nginx specialty that is nice,
but irrelevant AFAICS. ModSecurity is not involved in the
setup of the connection. Also, I think we defined to focus
on Apache / ModSec 2.9 first since it's the reference implementation.
That platforms allows you to write rules that report detailed
performance information. See 
https://www.netnea.com/cms/apache-tutorial-5_extending-access-log/
for metrics that you can implement in Apache. This allows you
to report total rule execution time per ModSec phase in the access log.
See tutorial 6 for the rules it takes to pull this off.

**Comment by dexter-ys** [comment]
@dune73 Thank you for addressing these issues! Let me try to explain some missing parts

> Not sure I responded on the performance hog question. The debug log can easily write 5-10K lines of log for a single GET request.

This is new to me. Maybe I should state with a specific log level here. In my test (from coreruleset/tests, log_level=`info`), one request will have:

```sh
# access.log (not used for the analysis)
192.168.16.1 - - [25/May/2023:18:06:49 +0000] "GET /status/200 HTTP/1.1" 200 - "-" "go-ftw test agent"
```

```sh
# error.log
[Thu May 25 18:06:49.595817 2023] [security2:error] [pid 49:tid 281473399632256] [client 192.168.16.1:65340] [client 192.168.16.1] ModSecurity: Warning. Pattern match "^.*$" at REQUEST_HEADERS:X-CRS-TEST. [file "/etc/modsecurity.d/owasp-crs/crs-setup.conf"] [line "757"] [id "999999"] [msg "60efed5b-b871-4571-8612-24f2278ccd26"] [tag "modsecurity"] [hostname "localhost"] [uri "/status/200"] [unique_id "ZG-jubXwTuATMgAAf3qkIQAAAIk"]
```

```sh
# modsec_audit.log
--64ea7071-A--
[25/May/2023:18:06:49.680508 +0000] ZG-jubXwTuATMgAAf3qkIQAAAIk 192.168.16.1 65340 192.168.16.3 80
--64ea7071-B--
GET /status/200 HTTP/1.1
Accept: */*
Connection: close
Host: localhost
User-Agent: go-ftw test agent
X-CRS-TEST: 60efed5b-b871-4571-8612-24f2278ccd26

--64ea7071-F--
HTTP/1.1 200 OK
Content-Type: text/html; charset=utf-8
Access-Control-Allow-Origin: *
Access-Control-Allow-Credentials: true
Content-Length: 0
Connection: close

--64ea7071-E--

--64ea7071-H--
Message: Warning. Pattern match "^.*$" at REQUEST_HEADERS:X-CRS-TEST. [file "/etc/modsecurity.d/owasp-crs/crs-setup.conf"] [line "757"] [id "999999"] [msg "60efed5b-b871-4571-8612-24f2278ccd26"] [tag "modsecurity"]
Apache-Error: [file "apache2_util.c"] [line 275] [level 3] [client 192.168.16.1] ModSecurity: Warning. Pattern match "^.*$" at REQUEST_HEADERS:X-CRS-TEST. [file "/etc/modsecurity.d/owasp-crs/crs-setup.conf"] [line "757"] [id "999999"] [msg "60efed5b-b871-4571-8612-24f2278ccd26"] [tag "modsecurity"] [hostname "localhost"] [uri "/status/200"] [unique_id "ZG-jubXwTuATMgAAf3qkIQAAAIk"]
Apache-Handler: proxy-server
Stopwatch: 1685038009594792 99675 (- - -)
Stopwatch2: 1685038009594792 99675; combined=20414, p1=18343, p2=1781, p3=38, p4=220, p5=32, sr=0, sw=0, l=0, gc=0
Response-Body-Transformed: Dechunked
Producer: ModSecurity for Apache/2.9.7 (http://www.modsecurity.org/); OWASP_CRS/4.0.0-rc1.
Server: Apache
Engine-Mode: "DETECTION_ONLY"

--64ea7071-Z--
``` 

---

> It's quite cool when dealing with ModSec 2.9 performance problems, but it's manual labor and the ModSec 3 debug log is less useful due to the absense of certain performance indicators inside the log.

Yes, I also noticed that some log options (e.g. PERF families) were removed. I will check on what information is in different types of logs. If error.log is all we want for evaluating specific matrices (e.g. RTT), we don't need debug.log.

---

> I would rather not use it, as it's very specific to various versions and I am sure it slows down a server.

May I know what is your concern that this approach will slow down the server?

My assumption of how people use the framework is that it works in the test environment instead of production, and we will try to simulate how production works (e.g. all configurations remained unchanged like log functions are open).

Meanwhile, My framework is based on separate server and performance testing utils (sidecar containers/script). So the server works independently without these sidecar items (e.g., a client is still available to send a request to the server directly via port 80). So the only change is that a user sends the request to fuzz proxy instead of WAF directly.

Yes, the resource will be consumed by additional sidecar containers, but it behaves on Docker instead of the server. Also, the performance evaluation is based on comparing before/after the server instead of evaluating one test result, it might slightly affect the server performance, but I think it is acceptable if we can measure how much performance is affected by this (or, restricting sidecar resource limit).

---

> Your results testing 932115-1 look interesting. I am not entirely sure I can make a call on this, but if you feel the subtle differences are enough and they are consistent, then I think you are on good tracks with one of these methods. What I think a bit troubling is that the two methods bring differring results. go-ftw got faster, the external script got slower. I think you need to make a substantial change to the regex. A mere negation is trivial for the engine and could lead you to run after a red herring.

Yes! This PoC was a week ago. I just randomly picked one that time. I will try different tests to see how it works. 

---

> Code: What is gdate? Gnu version of date? GO version of date?

gnu-date from [coreutils](https://formulae.brew.sh/formula/coreutils). `date` in MacOS does not support mili-sec. :(

---

> Other than that you make proficient use of standard shell tools and bash. This code looks good. Nice to see young people use awk and bc.

Appreciate that. Scripting is SUPER fun for me. 😆

---

> "Important thought: ..." - You might have gotten me wrong here. The problem is not that the server is not available. The problem is that - in my experience - in any HTTP transaction the client uses far more resources than the server. So even if everything was deterministic, subtle changes in the environment of the client have a huge effect that hide the performance differences on the server side.

In the framework, we use the same client to execute the same function (e.g., use a terminal to run a load test, or just run go-ftw) as a scientific constant. Server configuration is a scientific constant as well. The only change is the code change. Besides, the performance matrices (e.g., functionality, hardware) are collected from the server directly, whether from logs or docker.

---

> I like the idea around the sidecar architecture. I can not really comment on the container use. What is unclear to me is the role and need for cAdvisor when compared to ES / Prometheus. Are these options / alternatives?

Yes! They are optional. In 2nd meeting, we discussed possible solutions to retrieve hardware matrices. We can use cAdvisor (cannot export the result) or ElasticSearch/Prometheus (can export the result). This is still under discussion.

---

> Fuzzying: I see the need for diverse test cases, but I am not sure fuzzying of payloads is the right approach and if you get consistent results by introducing small changes into the payload via fuzzing.
> One problem with fuzzing I see (I think you saw it too) is the different rules hitting, thus changing the output (and therefore also changing the resources needed to write the output). Very hard to use the resulting numbers for calculations.

If we focus on testing one rule, fuzzing will be a good idea (since we know the scientific constant). Therefore, we can send the request to fuzz proxy.

Similarly, if we want to test all rules simultaneously, we can send the request to the WAF directly.

The Fuzz proxy is a sidecar container, which is optional, for a user to send their request, and the only functionality of it is to decorate the request (for testing convenience).

---

> Maybe it helps if you explain your reasoning for fuzzing. Maybe to Felipe in the call.

I added the fuzzing into the framework for design (as in 1st meeting mentioned that the framework is the foremost thing we want). It is first raised in this draft, and I will mention it in today's meeting for sure.

---

> Matrices: Don't use the modsec_audit.log. It's far too hard to parse and you do not need this level of information. Going with the error log is all you need (-> naked alerts without the rest of the request and response).

Thanks! Just checked the context in error.log again. Yes, it covers all the information we want to evaluate the error rate.

---

> Connection Time: This is an nginx specialty that is nice, but irrelevant AFAICS. ModSecurity is not involved in the setup of the connection. Also, I think we defined to focus on Apache / ModSec 2.9 first since it's the reference implementation. That platforms allows you to write rules that report detailed performance information. See https://www.netnea.com/cms/apache-tutorial-5_extending-access-log/ for metrics that you can implement in Apache. This allows you to report total rule execution time per ModSec phase in the access log. See tutorial 6 for the rules it takes to pull this off.

Agree. I just listed the possible matrices that we might be able to collect and waiting for further discussion. I will check on the link as well :)

---

Again, thanks for the feedback! 👍

**Comment by jcchavezs** [comment]
Sorry for being late to the party. As far as I understand you wanna meassure the impact of a rule in the critical path of a request. I think this is specially complicated depending on the environment. For example, in coraza you can run it as a middleware and hence the latency added by the rule, the memory consumption and the CPU usage is accurate but if you aim to run it through a proxy or a web server many things come to play: http latency, copy of the data (buffers for example), performance of the rule in a given language (e.g. native go for regexes vs re2 in c++ or wasm) so I think whatever approach that follows the blackbox design will end up in unnacurate results among WAFs.

In the other hand, if the only way to deploy the WAFs is through a proxy then it is probably not worth to compare them all and just do independent analysis because it isn't a fair comparison. What is more, concurrency can play a game here, it isn't the same analysing the performance of a rule with one request in a server vs concurrently run them (which will be more close to a real situation).

In sum I think making conclusions out of memory consumption, CPU usage and TTR **would only make sense if the results are consistent across WAFs** (which I don't know they are but I would discard outlier) and then we could come up with scales on how a rule performs and a [radar chart](https://en.wikipedia.org/wiki/Radar_chart) but I don't think this is useful to compare different WAFs.

**Comment by dexter-ys** [comment]
Hi @jcchavezs, thanks for the comments! 

> For example, in coraza you can run it as a middleware and hence the latency added by the rule, the memory consumption and the CPU usage is accurate but if you aim to run it through a proxy or a web server many things come to play: http latency, copy of the data (buffers for example), performance of the rule in a given language (e.g. native go for regexes vs re2 in c++ or wasm) so I think whatever approach that follows the blackbox design will end up in unnacurate results among WAFs.

Is it correct to interpret this as "measuring how long a rule is processed and acted on the server, excluding all the friction, such as network latency"?
If so, I am not sure if there is any way that we can measure the performance without modifying the source code so far based on the research. :(

In my mind, it seems to make sense to evaluate from a user's perspective - including the friction. (client-side matrices) and collect the matrices without being affected by that friction. (server-side matrices)

Right now, server-side matrices are proposed to collect from:
- functionality: from log (collected from server)
- hardware: from 3rd-party utils

and client-side matrices:
- functionality: from logs (on the client)

---

> In the other hand, if the only way to deploy the WAFs is through a proxy then it is probably not worth to compare them all and just do independent analysis because it isn't a fair comparison. What is more, concurrency can play a game here, it isn't the same analysing the performance of a rule with one request in a server vs concurrently run them (which will be more close to a real situation).

May I know how you typically do the independent analysis on a rule change? Meanwhile, I will check about the concurrency issue as well.

---

> In sum I think making conclusions out of memory consumption, CPU usage and TTR would only make sense if the results are consistent across WAFs (which I don't know they are but I would discard outlier) and then we could come up with scales on how a rule performs and a radar chart but I don't think this is useful to compare different WAFs.

The main performance evaluation is on one WAF and comparing before/after changes for a single rule. In other words, the evaluation of CPU/memory is compared for one rule change based on the same WAF. Nonetheless, I will also look into if there are any existing methodologies to compare cross-platform performance.

Thanks for the feedback!

**Comment by dexter-ys** [comment]
I will close this ticket while the prototype is developed. The latest document will be updated in the [README](https://github.com/dextermallo/GSoC-2023#readme). 

Thanks for all the feedback 🙌
